packages/labs/analyzer/src/lib/program-context.ts

```diff
@@ -0,0 +1,478 @@
+    return this.checker
+      .getSymbolsInScope(
+        location,
+        (ts.SymbolFlags as unknown as {All: number}).All
+      )
+      .filter((s) => s.name === name)[0];
+  }
+
+  /**
+   * Returns an analyzer `Type` object for the given jsDoc tag.
+   *
+   * Note, the tag type must
+   */
+  getTypeForJSDocTag(tag: ts.JSDocTag): Type {
+    if (!customJSDocTypeTags.has(tag.tagName.text)) {
+      throw new DiagnosticsError(
+        tag,
+        `Internal error: '${tag.tagName.text}' is not included in customJSDocTypeTags.`
+      );
+    }
+    const type = this.jsdocTypeMap.get(tag);
+    if (type === undefined) {
+      throw new DiagnosticsError(
+        tag,
+        `Internal error: no type was pre-processed for this JSDoc tag`
+      );
+    }
+    return this.convertType(type, tag);
+  }
+
+  /**
+   * Returns an analyzer `Type` object for the given AST node.
+   */
+  getTypeForNode(node: ts.Node): Type {
+    // Since getTypeAtLocation will return `any` for an untyped node, to support
+    // jsdoc @type for JS (TBD), we look at the jsdoc type first.
+    const jsdocType = ts.getJSDocType(node);
+    return this.convertType(
+      jsdocType
+        ? this.checker.getTypeFromTypeNode(jsdocType)
+        : this.checker.getTypeAtLocation(node),
+      node
+    );
+  }
+
+  /**
+   * Returns the module specifier for a declaration if it was imported,
+   * or `undefined` if the declaration was not imported.
+   */
+  getImportModuleSpecifier(declaration: ts.Node): string | undefined {
+    // TODO(kschaaf) support the various import syntaxes, e.g. `import {foo as bar} from 'baz'`
+    if (
+      ts.isImportSpecifier(declaration) &&
+      ts.isNamedImports(declaration.parent) &&
+      ts.isImportClause(declaration.parent.parent) &&
+      ts.isImportDeclaration(declaration.parent.parent.parent)
+    ) {
+      const module = declaration.parent.parent.parent.moduleSpecifier
+        .getText()
+        // Remove quotes
+        .slice(1, -1);
+      return module;
+    }
+    return undefined;
+  }
+
+  /**
+   * Returns an analyzer `Reference` object for a named identifier at the given
+   * AST location.
+   *
+   * If the symbol's declaration was imported, the Reference will be based on
+   * the import's module specifier; otherwise the Reference will point to the
+   * current module being analyzed.
+   */
+  getReferenceForSymbolName(name: string, location: ts.Node): Reference {
+    const symbol = this.getSymbolForName(name, location);
+    if (this.currentModule === undefined) {
+      throw new Error(`Internal error: expected currentModule to be set`);
+    }
+    const declaration = symbol?.declarations?.[0];
+    if (declaration === undefined) {
+      throw new DiagnosticsError(
+        location,
+        `Could not find declaration for symbol '${name}'`
+      );
+    }
+    if (declaration.getSourceFile() !== location.getSourceFile()) {
```

Comment:
If the symbol is imported, will the declaration still be in the location's source file? will it be an import declaration or something? Might make a good comment.

Reply:
So, I was seeing the declaration for an imported symbol being a `ImportSpecifier` (which is a `ts.Declaration` node) in the current file, not the declaration for wherever it was originally declared. However last night I was looking back at [your code here](https://github.com/lit/lit/blob/f8dad56b8e60b5e51510e22434031300b8e77f5a/packages/labs/analyzer/src/lib/lit-element/lit-element.ts#L41-L47), where the declaration for a symbol in the base type is coming from a different file, which made me slightly question that. 

I don't know whether it's just because `getBaseTypes` might return types with symbols resolved to their original declaration? The difference may be that this symbol is coming from `getSymbolsInScope`, which maybe has a `symbol.declaration[0]` that's the most local one?

Reply:
Ok, so looks like you get multiple declarations for:
* Classes (the interface + the constructor)
* Namespace merging (I think)
* ... but not automatically for the original declaration

It looks like you have to explicitly call `getAliasedSymbol()` to get the original declaration. So I think the current logic works, for now.

Added a good comment. 😉